R/zchunk_LA100.IMAGE_downscale_ctry_yr.R

```diff
@@ -30,61 +29,119 @@ module_aglu_LA100.IMAGE_downscale_ctry_yr_DISABLED <- function(command, ...) {
-    # then (iii) code a fix:
-    #
-    # if(OLD_DATA_SYSTEM_BEHAVIOR) {
-    #   ... code that replicates old, incorrect behavior
-    # } else {
-    #   ... new code with a fix
-    # }
-    #
-    #
-    # ===================================================
+    IMAGE_an_Feedfrac_Rimg_C_Sys_Fd_Y <- get_data(all_data, "aglu/IMAGE/IMAGE_an_Feedfrac_Rimg_C_Sys_Fd_Y")
+    IMAGE_an_FeedIO_Rimg_C_Sys_Y <- get_data(all_data, "aglu/IMAGE/IMAGE_an_FeedIO_Rimg_C_Sys_Y")
+    IMAGE_an_Prodmixfrac_Rimg_C_Y <- get_data(all_data, "aglu/IMAGE/IMAGE_an_Prodmixfrac_Rimg_C_Y")
+
+    # Helper function: copy 1970 data to new year 1960
+    extrapolate_1960 <- function(x) {
+      x %>%
+        filter(year == 1970) %>%
+        mutate(year = 1960) %>%
+        bind_rows(x)
+    }
+
+    # Extrapolate each IMAGE table to all historical years
+    # For each table, same basic steps: extrapolate to 1960; fill out all AGLU
+    # historical years; interpolate.
+
+    IMAGE_an_Feedfrac_Rimg_C_Sys_Fd_Y %>%
+      extrapolate_1960 %>%
+      gather(IMAGE_region_ID, value, -commodity, -system, -input, -year) %>%
+      complete(year = union(AGLU_HISTORICAL_YEARS, year), nesting(commodity, system, input, IMAGE_region_ID)) %>%
+      arrange(year) %>%
+      group_by(commodity, system, input, IMAGE_region_ID) %>%
+      mutate(value = approx_fun(year, value, rule = 2)) ->
+      L100.IMAGE_an_Feedfrac_Rimg_C_Sys_Fd_Y
+
+    IMAGE_an_FeedIO_Rimg_C_Sys_Y %>%
+      extrapolate_1960 %>%
+      gather(IMAGE_region_ID, value, -commodity, -system, -year) %>%
+      complete(year = union(AGLU_HISTORICAL_YEARS, year), nesting(commodity, system, IMAGE_region_ID)) %>%
+      arrange(year) %>%
+      group_by(commodity, system, IMAGE_region_ID) %>%
+      mutate(value = approx_fun(year, value, rule = 2)) %>%
+      ungroup %>%
+      # Re-set negative values in the feedfrac table to 0
+      mutate(value = if_else(value < 0, 0, value)) ->
+      L100.IMAGE_an_FeedIO_Rimg_C_Sys_Y
+
+    IMAGE_an_Prodmixfrac_Rimg_C_Y %>%
+      extrapolate_1960 %>%
+      gather(IMAGE_region_ID, value, -commodity, -year) %>%
+      complete(year = union(AGLU_HISTORICAL_YEARS, year), nesting(commodity, IMAGE_region_ID)) %>%
+      arrange(year) %>%
+      group_by(commodity, IMAGE_region_ID) %>%
+      mutate(value = approx_fun(year, value, rule = 2)) ->
+      L100.IMAGE_an_Prodmixfrac_Rimg_C_Y
+
+
+    # Downscale IMAGE region-level data to all countries
+
+    # Change AGLU_ctry$IMAGE_region_ID to character for easy joining in helper function
+    AGLU_ctry$IMAGE_region_ID <- as.character(AGLU_ctry$IMAGE_region_ID)
+
+    # Helper function: filter, repeat and add columns, join with iso data
+    downscale_IMAGE_regions <- function(x, AGLU_ctry, by) {
+      hyd <- filter(x, year %in% AGLU_HISTORICAL_YEARS) %>% spread(year, value)
+
+      x %>%
+        ungroup %>%
+        filter(IMAGE_region_ID == 1, ! year %in% AGLU_HISTORICAL_YEARS) %>%
```

Comment:
OK just saw this (I swear this was hidden when I looked an hour ago...). There's no significance of region 1 (which for IMAGE 2.4 is Canada, not the US); it could have been any region.